docs/developers_guide/FlinkComponents.md

```diff
@@ -1,22 +1,72 @@
 # Flink components
- 
-Sources, sinks and custom transformations are based on 
-Flink API. Access to various helpers is provided by 
-`FlinkCustomNodeContext`. Special care should be taken to handle
-- lifecycle
-- exception handling properly
+
+Sources, sinks and custom transformations are based on
+Flink API. Access to various helpers is provided by
+`FlinkCustomNodeContext`. Special care should be taken to handle:
+- lifecycle - preparing the operator (source, sink or functions registered by custom transformers), closing resources, handling failures and restoring the state.
+  See Flink's operators lifecycle for more details [here](https://ci.apache.org/projects/flink/flink-docs-master/docs/internals/task_lifecycle/)
+- exceptions, e.g. during deserialization, since any thrown and unhandled exception by the source, causes the Flink job to restart.
 
 :warning: **Flink components should not extend Lifecycle** - it won't be handled properly
 
 ## Sources
 
-Source are defined with [FlinkSource](https://github.com/TouK/nussknacker/blob/staging/engine/flink/components-api/src/main/scala/pl/touk/nussknacker/engine/flink/api/process/FlinkSource.scala).
-In most cases (when you only pass one variable to initial `Context`) it's easier to use `BaseFlinkSource`.
+Sources implementations are defined with [FlinkSource](https://github.com/TouK/nussknacker/blob/staging/engine/flink/components-api/src/main/scala/pl/touk/nussknacker/engine/flink/api/process/FlinkSource.scala).
+In most cases (when you only pass one variable to initial `Context`) it's easier to use `BasicFlinkSource`. This trait
+requires only Flink's `SourceFunction` to be implemented, whereas `FlinkSource` allows to prepare an arbitrary `DataStream`
+and add custom [context initializer](https://github.com/TouK/nussknacker/blob/staging/components-api/src/main/scala/pl/touk/nussknacker/engine/api/process/ContextInitializer.scala).
+You should bear in mind to apply a watermark strategy to the stream so that events will be processed correctly downstream, e.g. in aggregates.
+
+Sources usually emit a single variable named `#input` having some attributes. To have those attributes recognized
+by the designer you need to specify them. For more information consult [components specification](Components.md#specification)
+and [types](Basics.md#types) to describe (available fields, their types) the emitted `#input`.
+
+### Test data generation
+
+Sources optionally support generating test data to ease [scenarios development](../scenarios_authoring/TestingAndDebugging.md).
+In order to support this feature, `FlinkSourceTestSupport` trait has to be implemented by the source.
+
+### Integrating with Flink sources
```

Comment:
What is the difference (in terms of work to be done) of starting (?) from Flink source vs approach described in Sources section? What are advantages / disadvantages?